CSP/rules.py

```diff
@@ -0,0 +1,92 @@
+                pitch_variable = names[voice_index] + str(i + 1)
+                for j in [x for x in range(0, 4) if x != voice_index]:
+                    other_pitch = voices[j][i]
+                    other_pitch_next = voices[j][i + 1]
+                    if other_pitch == '??':
+                        other_pitch_variable = names[j] + str(i + 1)
+                        if pitch_next == '??':
+                            pitch_next_variable = names[voice_index] + str(i + 2)
+                            if other_pitch_next == '??':
+                                other_pitch_next_variable = names[j] + str(i + 2)
+                                problem.addConstraint(
+                                    lambda x1, x2, y1, y2:
+                                    not (abs(x1 - y1) % 12 == interval and interval == abs(x2 - y2) % 12 and x1 != x2),
+                                    [pitch_variable, pitch_next_variable, other_pitch_variable, other_pitch_next_variable])
+                    else:
+                        if pitch_next == '??':
+                            pitch_next_variable = names[voice_index] + str(i + 2)
+                            if other_pitch_next == '??':
+                                other_pitch_next_variable = names[j] + str(i + 2)
+                                problem.addConstraint(
+                                    lambda x1, x2, y2, y1=other_pitch:
+                                    not (abs(x1 - y1) % 12 == interval and interval == abs(x2 - y2) % 12 and x1 != x2),
+                                    [pitch_variable, pitch_next_variable, other_pitch_next_variable])
+                            else:
+                                problem.addConstraint(
+                                    lambda x1, x2, y1=other_pitch, y2=other_pitch_next:
+                                    not (abs(x1 - y1) % 12 == interval and interval == abs(x2 - y2) % 12 and x1 != x2),
+                                    [pitch_variable, pitch_next_variable])
+                # TODO: Implement extended cases
+            voice_index += 1
+
+
+def seventh_resolution(voices, harmony, problem, names):
+    for i in range(0, len(harmony)):
+        if harmony[i][1] in ['7', 'D', 'd']:
+            pitches = expand_harmony(harmony[i])
+            voice_index = 0
+            for voice in voices:
+                if voice[i] == '??':
+                    pitch_variable = names[voice_index] + str(i + 1)
+                    if voice[i + 1] == '??':
+                        pitch_next_variable = names[voice_index] + str(i + 2)
+                        problem.addConstraint(lambda x1, x2, ref=pitches[3]:
+                                              True if x1 % 12 == ref and (x1 - x2 == 1 or x1 - x2 == 2) else
+                                              (True if x1 % 12 != ref else False), [pitch_variable, pitch_next_variable])
+                voice_index += 1
+                # TODO: Implement the rest of the cases
```

Comment:
Again, is this a later potential improvement of the project?

Reply:
Indeed